Allow the OEM to control what the ODM is able to do

hughsie opened the pull request:
To enforce this, we also need to add the concept of vendor namespaces.
The namespace prefix is the AppStream ID prefix that can be displayed to the
user. The OEM is the 'owner' of the ID prefix, and never the ODM. Add a
firmware 'problems' for existing firmware on the LVFS that hasn't already been
pushed to stable.

This prevents the ODM being 'credited' with the hardware when they are just
managing the entire end-to-end firmware lifecycle for the OEM as part of a
contract. Adding the actions to an ACL list allows us to better explain to the
vendor manager what the ODM is able to do, without hardcoding the logic.

Additionally, add an optional notfication email for ODM or OEM upload so that
the OEM is aware of what the ODM is doing.

Fixes https://github.com/fwupd/lvfs-website/issues/380

hughsie commented:
![Screenshot_2019-09-04 LVFS Vendor Affiliations(1)](https://user-images.githubusercontent.com/151380/64295946-95e7f300-cf6a-11e9-92b3-1362d948135d.png)

hughsie commented:
![Screenshot_2019-09-04 LVFS Vendor Affiliations(2)](https://user-images.githubusercontent.com/151380/64296055-e8291400-cf6a-11e9-90d9-7cf000de64e6.png)

superm1 commented on lvfs/models.py:
```diff
@@ -1369,6 +1433,23 @@ def problems(self):
+                              'Component ID {} not allowed '
+                              'for vendor {}: {}'.format(self.appstream_id_prefix,
+                                                         self.fw.vendor_odm.group_id,
+                                                         ','.join(values)))
```
I just want to make sure i'm following this correctly.  So basically in the XML say you have `com.dell` as your appstream prefix and Compal uploads some firmware.  As long as Compal has `com.dell` in their vendor_odm.namespaces they'll be able to do the actions that have been granted to them, correct?

hughsie replied:
>  As long as Compal has com.dell in their vendor_odm.namespaces they'll be able to do the actions

Not quite. What this is saying is that only Dell would have the com.dell id prefix, so the firmware vendor has to be assigned to dell and **never** to compal. This means we'd never refer to the firmware as a "Compal XPS 13" device which could happen in theory now if compal pushed to stable without changing the vendor to dell beforehand. What this patch does is force the ODMs to assign the correct OEM as the vendor, and then uses the user_id which is the uploader to work out the vendor group of the uploader, which is used for the odm fallback. I think it's 100x more clear than the old version where you had three things, the vendor_oem, vendor_odm and the user_uploader.

superm1 replied:
OK, the most important thing to me was making sure that Compal didn't have to change the XML in some way, it's literally just a form upload action and database matching then.

hughsie replied:
Right, on the upload form they'd just chose Dell rather than Dell or Compal. If they chose Compal they'd get a "problem" appear, and then be able to transfer the firmware to Dell to remove the problem.

superm1 commented on lvfs/models.py:
```diff
@@ -1791,10 +1874,32 @@ def _ensure_blobs(self):
         self.is_dirty = True
         self.remote.is_dirty = True
 
+    def _is_vendor_or_permitted_odm_action(self, action, user):
```
just a nit, how about maybe `is_permitted_action` and then leave the details to why it's permitted to the function rather than the title?

hughsie replied:
You mean just rename the function, or move the `_is_vendor()` check back to the `if` in the caller?

superm1 replied:
Just personal preference was leaning on renaming it rather than putting the logic in the title.

hughsie replied:
Done.

superm1 commented on lvfs/templates/email-firmware-uploaded.txt:
```diff
@@ -0,0 +1,15 @@
+You can log into the LVFS and view the firmware: {{url_for('.firmware_analytics_reports', firmware_id=fw.firmware_id, state=3, _external=True)}}
+
+If you believe this action should not have been allowed please contact the LVFS administrator by forwarding the email to info@fwupd.org with details.
+
```
I think this template is great.  I think that a similar template for promotion/demotion actions would be quite interesting too though.

hughsie replied:
Agreed. The User table is becoming a bit of a huge boolean mess so I thought I would clean that up at some stage, possibly with a 1:1 helper table just for norifications, or perhaps switch them to string enums so I don't keep having to do db changes when adding email notification types.

superm1 commented:
From what I follow this looks fine to me.  The various permissions that exist, I assume those will all default to not enabled and then will just turn on the individual ones as needed.

hughsie commented:
> I assume those will all default to not enabled and then will just turn on the individual ones as needed.

For new affiliations, exactly that. For existing affiliations I used: `@add-limit,@delete,@modify,@undelete,@modify-updateinfo,@view` -- is that sane?

Also, do you think vendor managers should be able to control this themselves, or leave this to the admin users?

superm1 commented:
> add-limit

I don't think this should default for existing ones.  If there is a company using it, they should share it with their affiliations only.

>remove-limit

I think that you may just want to merge these to `modify-limit` so you don't accidentally end up in a situation that an ODM turns on a limit while twisting knobs to see what they do and can't turn it back off.

> modify-keywords, modify-requirements, modify-checksums

I'd think this should default to yes for existing.

>Also, do you think vendor managers should be able to control this themselves, or leave this to the admin users?

Well affiliation has to be set by admin initially right?  I think that vendor manager should be able to at least see read-only see affiliation permission details, but don't see a problem with them adjusting the write permissions too.

hughsie commented:
>> add-limit
> I don't think this should default for existing ones

Agreed, removed.

> I think that you may just want to merge these to modify-limit

Done.

> Well affiliation has to be set by admin initially right

Yes, as the ODM might not be "public" on the LVFS yet, if ever. The vendors using the LVFS is kinda secret at this point.

> but don't see a problem with them adjusting the write permissions too.

I'll work on that this morning.

hughsie commented:
> I'll work on that this morning.

Done, I think.

superm1 commented:
LGTM now.

hughsie commented:
I was thinking of sending something like this to lvfs-announce, ideas welcome:

Planned new functionality for managing the ODM and OEM relationship

At the moment the LVFS has lots of hardcoded assumptions that the ODM uploads the firmware, performs some testing in embargo and then transfers ownership to the OEM which add update details, and moves the firmware to testing and stable. This doesn’t match reality for a number of large vendors on the LVFS (for example, where the ODM manages all parts of the update flow), and so we needed to do something better going forward. There is also the problem where the ODM is being “credited” with the firmware upload, when we really don’t want to leak out the ODM→OEM relationship even accidentally.

To do this I’ve changed slightly how we assign the firmware on upload. Each OEM has a set of appstream prefixes that match the <id> values uploaded to the LVFS. The idea is that, for instance, Compal would never own firmware with com.dell.* and only the Dell account can do that. This would stop the ODM accidentally owning firmware that should be credited to the OEM.

The user who uploaded the firmware is part of a vendor group, and this is how we get the ODM vendor information. If an affiliation has been defined between the OEM and ODM then the ODM can manage the firmware on the LVFS. We can now define what “manage” means, with the most simple permissions being “just view” and the most complete permissions being “view, modify, delete, move to testing, move to stable”.

For all the vendors on the LVFS I will assign what I think are the correct vendor prefixes, based on the existing values uploaded, and for vendors with no uploaded firmware a prefixed based on the reverse DNS of the OEM homepage. For instance, the Dell account will have ‘com.dell’ assigned. OEMs can have multiple vendor prefixes assigned where required, and the vendor prefix doesn’t have to match the domain name in every case.

There are a lot of existing ODM<->OEM vendor relationships on the LVFS. I’ve converted the existing relationships to be a safe subset of the existing permissions on the affiliate relationship. For vendors who are currently managing the end-to-end flow the OEM vendor manager will need to add extra allowed actions for the ODM. Either ask your OEM vendor manager to do this, or me.

I’m aware this changes the way some ODMs are using the LVFS right now, and I’m happy to help both OEMs and ODMs choose actions that reflects the current business relationship. There might be firmware files that I’ve not fixed, or where affiliations need to be added or modified, and please email me if you’re suddenly unable to do an action that you expected to be able to do.

Existing firmware in testing and stable isn’t affected by the new rules, and this will only affect new firmware uploaded. If anyone has any concerns please say. I’m confident the new functionality will allow us to grow the LVFS and also be suitable for the changing landscape of ODM and OEM relationsips.

Richard.

superm1 commented:
> This doesn’t match reality for a number of large vendors on the LVFS (for example, where the ODM manages all parts of the update flow), and so we needed to do something better going forward

This doesn't match the _desired_ flow.  I'd think that no ODMs should be able to promote OEM firmware to stable until this.

> To do this I’ve changed slightly how we assign the firmware on upload. Each OEM has a set of appstream prefixes

I would highly suspect people who receive this message don't know what appstream prefixes are.  If there is a link you can send (or document on LVFS) what they are I think it would be useful.

> For all the vendors on the LVFS I will assign what I think are the correct vendor prefixes, based on the existing values uploaded, and for vendors with no uploaded firmware a prefixed based on the reverse DNS of the OEM homepage

No need to mention what you "think are the correct vendor prefixes".  Maybe instead: "I will assign vendor prefixes based upon existing values uploaded"

> Existing firmware in testing and stable isn’t affected by the new rules, and this will only affect new firmware uploaded. If anyone has any concerns please say. I’m confident the new functionality will allow us to grow the LVFS and also be suitable for the changing landscape of ODM and OEM relationsips.

I think there will be questions about stuff _currently_ in $oem-embargo.  Can ODM promote those all the way to stable now given the right permissions?

hughsie commented:
> No need to mention what you "think are the correct vendor prefixes". Maybe instead: "I will assign vendor prefixes based upon existing values uploaded"

Right, thanks.
 
> I think there will be questions about stuff _currently_ in $oem-embargo. Can ODM promote those all the way to stable now given the right permissions?

Good question. Either I (probably it'll be me...) or the vendor manager will have to reassign these before pushing. I'll obviously not deploy this until I've heard back from the other two bit OEMs using the affiliates feature (so the ODMs get the right permissions set from day 0) and also when I can view all the firmware problems and fix them manually. I also need to rerun several months of stats calculations with the new vendor's set correctly. (e.g., Wistron will disappear from the graph)
